### PR TITLE
Use `subprocess` for checking XLA supported devices.

### PR DESCRIPTION
**Pull request by ysiraichi**
Fix: #5943

In summary:

- Create a self-contained XLA device checking python script: check_xla_device.py
- Use `subprocess` for calling the checking script with different devices

### CODE REVIEW

**Comment by ysiraichi** [comment]
cc @JackCaoG @miladm

File: benchmarks/util.py [hunk 1]
```diff
@@ -51,28 +49,10 @@ def reset_rng_state(benchmark_experiment=None):
...
-  process.start()
-  process.join(60)
-  try:
-    return q.get_nowait()
-  except queue.Empty:
-    traceback.print_exc()
-    return False
+  CHECK_XLA_DEVICE_PY = "check_xla_device.py"
+  python_file = os.path.join(os.path.dirname(__file__), CHECK_XLA_DEVICE_PY)
+  r = subprocess.run([sys.executable, python_file, devkind])
```
Comment (frgossen) [inline benchmarks/util.py:54]: Why does this fix the issue? (I never understood why this failure appeared in the first place) 
I just want to make sure that we don't hide the error message in a subprocess and silently fail.

**Review by frgossen — CHANGES_REQUESTED** [review]

**Comment by vanbasten23** [comment]
I ran your repro https://github.com/pytorch/xla/issues/5943#issuecomment-1834697201 and I got:
```
root@xiowei-gpu:/ansible# PJRT_DEVICE=CUDA python test.py
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1701747812.339996    9300 tfrt_cpu_pjrt_client.cc:349] TfrtCpuClient created.
R: True
R: True
root@xiowei-gpu:/ansible#
```
Is it consistent with what you got?

**Comment by ysiraichi** [comment]
No. I still get:

```
$ PJRT_DEVICE=CUDA python repro.py

WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1701790031.458193  396674 tfrt_cpu_pjrt_client.cc:349] TfrtCpuClient created.
R: True
2023-12-05 15:27:11.514993: E external/xla/xla/stream_executor/cuda/cuda_driver.cc:274] failed call to cuInit: CUDA_ERROR_NOT_INITIALIZED: initialization error
Traceback (most recent call last):
  File "repro.py", line 10, in _check
    q.put(bool(xm.get_xla_supported_devices(devkind=k)))
  File "xla/torch_xla/core/xla_model.py", line 99, in get_xla_supported_devices
    xla_devices = _DEVICES.value
  File "xla/torch_xla/utils/utils.py", line 29, in value
    self._value = self._gen_fn()
  File "xla/torch_xla/core/xla_model.py", line 20, in <lambda>
    _DEVICES = xu.LazyProperty(lambda: torch_xla._XLAC._xla_get_devices())
RuntimeError: Bad StatusOr access: FAILED_PRECONDITION: No visible GPU devices.
R: False
```

**Comment by ysiraichi** [comment]
Maybe that has something to do with the underlying OS. Using `spawn` start method (default on some systems, I believe) does fix the issue. However I couldn't make it work with the benchmarking script.

**Comment by frgossen** [comment]
> No. I still get:
> 
> ```
> $ PJRT_DEVICE=CUDA python repro.py
> 
> WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
> I0000 00:00:1701790031.458193  396674 tfrt_cpu_pjrt_client.cc:349] TfrtCpuClient created.
> R: True
> 2023-12-05 15:27:11.514993: E external/xla/xla/stream_executor/cuda/cuda_driver.cc:274] failed call to cuInit: CUDA_ERROR_NOT_INITIALIZED: initialization error
> Traceback (most recent call last):
>   File "repro.py", line 10, in _check
>     q.put(bool(xm.get_xla_supported_devices(devkind=k)))
>   File "xla/torch_xla/core/xla_model.py", line 99, in get_xla_supported_devices
>     xla_devices = _DEVICES.value
>   File "xla/torch_xla/utils/utils.py", line 29, in value
>     self._value = self._gen_fn()
>   File "xla/torch_xla/core/xla_model.py", line 20, in <lambda>
>     _DEVICES = xu.LazyProperty(lambda: torch_xla._XLAC._xla_get_devices())
> RuntimeError: Bad StatusOr access: FAILED_PRECONDITION: No visible GPU devices.
> R: False
> ```

FWIW, I get this about every second time I run the script.

**Comment by vanbasten23** [comment]
@ysiraichi @frgossen iirc, can you check one thing? Have your env var PATH and LD_LIBRARY_PATH accounted for cuda, for example https://github.com/pytorch/xla/blob/master/docs/gpu.md#check-environment-variable?

**Comment by ysiraichi** [comment]
For some reason, the same script started working for me (even without setting the environment variables). This is all looking very flaky.

**Comment by ysiraichi** [comment]
That said, here's another reproducer:

```python
import traceback
import torch_xla.core.xla_model as xm

from multiprocessing import Process, Queue

import torch
print(torch.cuda.is_available())

def _check(q, k):
    import os
    os.environ["PJRT_DEVICE"] = k
    try:
        q.put(bool(xm.get_xla_supported_devices(devkind=k)))
    except:
        traceback.print_exc()
        q.put(False)

for devkind in ("CPU", "CUDA"):
    q = Queue()
    proc = Process(target=_check, args=(q, devkind))
    proc.start()
    proc.join(60)
    print("R:", q.get_nowait())
```

**Comment by ysiraichi** [comment]
Apparently, this is a known limitation: you can't initialize a CUDA context before forking the process (and use the initialized context in the children process).

**Comment by frgossen** [comment]
I have the env vars set and can confirm that it is very flaky for me.

**Comment by vanbasten23** [comment]
The change lgtm. @ysiraichi , could you add a comment about why we need to do this way?

**Review by vanbasten23 — APPROVED** [review]

**Comment by vanbasten23** [comment]
The CI build fails. Let me retrigger it

**Comment by ysiraichi** [comment]
Yes. I had some PRs also fail after building successfully (operation cancelled, etc.). Pretty sure it's unrelated to the PR. Any thoughts?

**Comment by ysiraichi** [comment]
Closed due to [a timeout issue](https://github.com/pytorch/xla/pull/6061#issuecomment-1847632833). Re-opened the PR at #6068.

**Pull request closed without merging**